Models: Add ChangeVIT model

hkristen opened the pull request:
Cleanroom implementation of the [ChangeViT paper](https://arxiv.org/pdf/2406.12847). The [code shared](https://github.com/zhuduowang/ChangeViT) by the authors is CC-BY-NC-SA-4.0 **therefore we can't copy from there or even look at the code while implementing.** 

This initial version should follow the paper quite closely. It has a some tests, a few of them where written at the beginning to make sure I don't go to far off while implementing. I am sure I over-complicated things, very happy for suggestions to make it more lean!

### TODO:
- [x] ~Tolerance paramter size ok? -> timm vit + standard resnet vs. custom VIT + custom resnet (I think something is off at the decoder)~
- [x] ~Which weights should be used as standard? Should we add DINO-S weights as well?~
- [x] Go through code once more and make sure it aligns with paper
- [x] Train model on benchmark datasets (Levir-CD, WHU-CD) and compare with paper results
- [x] Split out the datamodule/transforms stuff to #3028 
- [x]  after #3028 is merged, rebase changevit branch and merge to main

---- 

Now I want to run a benchmark training run with the same parameters as in the paper to ensure the implementation produces similar results on LEVIR-CD (which is the actual test if my implementation makes sense). I am using the CLI for it but get this error when I try to run it and can't figure out what the issue is:

`torchgeo fit --config tests/conf/changevit_levircd_benchmark.yaml `
```
error: Parser key "model":
  Does not validate against any of the Union subtypes
  Subtypes: [<class 'NoneType'>, <class 'torchgeo.trainers.base.BaseTask'>]
  Errors:
    - Expected a <class 'NoneType'>
    - Problem with given class_path 'torchgeo.trainers.ChangeDetectionTask':
        Parser key "model":
          Cannot take a Union of no types.
  Given value type: <class 'dict'>
  Given value: {'class_path': 'ChangeDetectionTask', 'init_args': {'model': 'changevit_tiny', 'task': 'binary', 'loss': 'bce', 'lr': 0.0002, 'num_classes': 2}}
```
These seems like a dumb error - any suggestions @adamjstewart what the issue is here?

hkristen commented on torchgeo/trainers/change.py:
```diff
@@ -281,7 +344,15 @@ def _shared_step(self, batch: Any, batch_idx: int, stage: str) -> Tensor:
         metrics = getattr(self, f'{stage}_metrics', None)
         if metrics:
-            metrics(y_hat, y)
+            # Transform predictions for metrics calculation (don't modify original y_hat)
```
This is not needed anymore. Was a fix I implemented while developing and forgot about it. Cleaned this and some other up now

blaz-r commented on torchgeo/trainers/change.py:
```diff
@@ -263,13 +294,20 @@ def _shared_step(self, batch: Any, batch_idx: int, stage: str) -> Tensor:
         y = batch['mask']
 
-        if not model.startswith('fcsiam'):
+        if not model.startswith('fcsiam') and not model.startswith('changevit'):
             x = rearrange(x, 'b t c h w -> b (t c) h w')
```
This might also be needed for other models, including BTC. Right now, I just do the rearrangement in the forward, but maybe we could think of a way to handle this better?

hkristen replied:
Sure, I just added changevit to the already existing fcsiam. Feel free to add more where it is needed.

blaz-r commented on torchgeo/trainers/change.py:
```diff
@@ -373,11 +412,22 @@ def predict_step(
         if model == 'unet':
             x = rearrange(x, 'b t c h w -> b (t c) h w')
-        y_hat: Tensor = self(x)
+        elif not model.startswith('fcsiam') and not model.startswith('changevit'):
+            x = rearrange(x, 'b t c h w -> b (t c) h w')
```
Same here.

blaz-r commented on torchgeo/datamodules/levircd.py:
```diff
@@ -1,150 +1,174 @@
+        return K.AugmentationSequential(
+            K.Normalize(mean=self.mean, std=self.std),
+            _ExtractPatches(
+                window_size=self.patch_size, stride=self.stride, keepdim=False
```
The updated levir and transforms don't quite work for me. I think that this should be set to `keepdim=True`, so that the correct reshaping triggers.

blaz-r commented on torchgeo/transforms/transforms.py:
```diff
@@ -87,8 +204,25 @@ def apply_transform(
+        if (
+            len(out.shape) == 5
+            and out.shape[1] > 1
+            and input.shape[0] % 2 == 0
```
I don't think this check here is a good way to test for temporal data. It then fails on masks since they have shape of 1.

hkristen commented:
> The validation crops work fine, the training one doesn't, I think. If I also use the extract patches in training, then it starts working.

you are right, i only checked the validation patches 🤦 here is one example from training:
<img width="2312" height="595" alt="grafik" src="https://github.com/user-attachments/assets/a3957bcc-92d5-45b0-a1a0-4064bb72fcc1" />

hkristen commented:
I had another look, changed back to using Kornia and my training now works. Please verify @blaz-r 

@adamjstewart we would need to move the changes in those files to a new PR and then after is merged continue with the ChangeViT PR as some tests depend on it:
```
torchgeo/datamodules/__init__.py
tests/conf/levircdplus.yaml
torchgeo/datamodules/__init__.py
torchgeo/datamodules/levircd.py
torchgeo/transforms/transforms.py
```

hkristen commented:
Thanks @calebrob6  - I did one training run before my holiday, which went well but didn't fully re-produce the paper results. I will go through the latest changes proposed by @isaaccorley next week and then do another training run. Should be closer then I guess, will report back here.

hkristen commented:
Status:
I re-based to main as I didn't work on this for quite some time and addressed all comments from @isaaccorley 

Next steps are:
- [ ]  Train model on benchmark datasets (Levir-CD, WHU-CD) and compare with paper results
- [ ] split out PR for changes regarding datamodule/transform stuff 
- [ ] after the datamodule/transform PR is merged, the changevit branch can be merged to main

hkristen commented:
Training is running well, I get quite good training metrics but the validation scores for F1 and Jaccard are abysmal when I use the hyperparamters reported in the paper. I tried a few different loss functions and learning rates over the weekend but the result is very similar. 

<img width="310" height="402" alt="image" src="https://github.com/user-attachments/assets/17ebbada-0fb5-4498-9676-7e715dc297a8" />
<img width="1626" height="539" alt="Screenshot from 2025-10-06 12-53-35" src="https://github.com/user-attachments/assets/6b43ae02-c3a2-4044-a6f9-620b6f70fda5" />
<img width="1626" height="539" alt="Screenshot from 2025-10-06 12-54-02" src="https://github.com/user-attachments/assets/398daf58-408f-403c-936e-70c115caaf26" />

I suspect that the issue is still somewhere in the validation augmentation of the Levir-CD dataset but I am kinda stuck. Can someone else try and have a look? @isaaccorley or @calebrob6 maybe?

isaaccorley commented on torchgeo/trainers/change.py:
```diff
@@ -298,6 +358,7 @@ def _shared_step(self, batch: Any, batch_idx: int, stage: str) -> Tensor:
                 and hasattr(self.logger.experiment, 'add_figure')
             ):
                 datamodule = self.trainer.datamodule
+
```
```suggestion
```

isaaccorley commented on torchgeo/trainers/change.py:
```diff
@@ -267,12 +325,14 @@ def _shared_step(self, batch: Any, batch_idx: int, stage: str) -> Tensor:
             y = y.squeeze(1)
+            y = y.long()
 
+        # Forward pass
```
```suggestion
```

isaaccorley commented on tests/models/test_changevit.py:
```diff
@@ -0,0 +1,530 @@
+        for comp_name in components_to_check:
+            component = getattr(model, comp_name)
+            has_grad = any(p.grad is not None for p in component.parameters())
+            assert has_grad, f'Component {comp_name} should receive gradients'
```
```suggestion
```

I don't think these tests need to run in CI on every commit

isaaccorley reviewed (changes requested):
A few thoughts on this PR:

- model tests should be redone to match the style of other model tests. consider that each test added adds more time to CI so we should try to minimize the number of tests that will actually reach 100% coverage.

- trainer tests should go in the existing change detection trainer test scripts

- because we don't have any actual pretrained weights for this model, we should not need a WeightsEnum and this should be written similarly to FarSeg and ChangeStar https://github.com/torchgeo/torchgeo/blob/main/torchgeo/models/changestar.py
- because the model size is dictated by the timm backbone, it seems unecessary to have separate `changevit_XXX` functions e.g. `changevit_small`. I would recommend just maintaining a single `changevit` function.

hkristen commented:
Thanks for your comments @isaaccorley - I will factor those in tomorrow. 

One more thing: I did start a training run with the config `tests/conf/changevit_levircd_dinov3_large_sat_optimized.yaml` to see what effect the changes made by @adamjstewart have on the training metrics. They are quite bad now, see screenshot below (Purple = before changes from main where merged, Turqouise = current state of the ChangeViT branch)
<img width="1253" height="260" alt="image" src="https://github.com/user-attachments/assets/eadf27af-ac80-48f3-ab73-381ff2628dad" />

I suspect that changing back to dice loss vs. dice+bce before and using only centercrop are the culprits. I get it, why those changes where made but I don't think it is ideal that paper results can't be reproduced with the current state of the model & levircd dataset. 

My question is how to proceed from here? As I guess this will partly be resolved by the outcomes of https://github.com/torchgeo/torchgeo/issues/3108.

calebrob6 commented:
Few things:

@hkristen -- heroic effort here re-implementing and verifying from scratch, that's awesome!!! I'm really excited to have changevit in torchgeo. Nice job!

Do you mean https://github.com/torchgeo/torchgeo/blob/7ba3c692c07d78a425adced27ff7fa841c211bc1/torchgeo/datamodules/levircd.py#L64 is messing up your metrics? I agree that only evaluating test metrics over center crops because the images are rectangular isn't the way to go.

hkristen commented:
Okidoki - I tried to address all review comments from @isaaccorley and started a full overnight training run. Let's see how it goes. 

Let's figure out the rest later

adamjstewart commented on tests/trainers/test_change.py:
```diff
@@ -242,6 +269,38 @@ def test_btc_model(self) -> None:
+        y = model(x)
+        assert y.shape == (1, 5, 256, 256)
+
+    def test_changevit_predict(self) -> None:
```
Technically none of these are needed, but I'll simplify the tests another day.

adamjstewart reviewed (dismissed):
LGTM with the following caveats. Would like to hear from @hkristen before merging to make sure I didn't break anything, but I think we have 100% test coverage finally.

isaaccorley commented on tests/models/test_changevit.py:
```diff
@@ -0,0 +1,116 @@
+        assert c4.shape[-2:] == (32, 32)  # 256 / 8
+
+
+class TestFeatureInjector:
```
nit: Same as above comment

adamjstewart reviewed (approved):
Let's merge, hopefully any improvements needed will be backwards-compatible.

hkristen commented on torchgeo/trainers/change.py:
Works for me, consistent behavior across all models is cleaner.
